internal/revision/database/revision.go

```diff
@@ -0,0 +1,292 @@
+}
+
+// GetAllowedRevisionKeys returns all of the curently allowed revision keys.
+// This method will unwrap all of the keys so that they can be used to create and verify
+// revision tokens.
+func (rdb *RevisionDB) GetAllowedRevisionKeys(ctx context.Context) (int64, []*RevisionKey, error) {
+	logger := logging.FromContext(ctx)
+	var effectiveID int64
+	keys := make([]*RevisionKey, 0)
+	if err := rdb.db.InTx(ctx, pgx.ReadCommitted, func(tx pgx.Tx) error {
+		rows, err := tx.Query(ctx, `
+			SELECT
+				kid, aad, wrapped_cipher, created_at, allowed
+			FROM
+				revisionkeys
+			WHERE
+				allowed=$1
+			ORDER BY created_at DESC`, true)
+		if err != nil {
+			return nil
+		}
+		defer rows.Close()
+
+		for rows.Next() {
+			if err := rows.Err(); err != nil {
+				return fmt.Errorf("failed to iterate: %w", err)
+			}
+			var r RevisionKey
+			if err := rows.Scan(&r.KeyID, &r.AAD, &r.WrappedCipher, &r.CreatedAt, &r.Allowed); err != nil {
+				return err
+			}
+			keys = append(keys, &r)
+			// The effective KEY is the first one due to ORDER BY clause.
+			if effectiveID == 0 {
+				effectiveID = r.KeyID
+			}
+		}
+
+		return nil
+	}); err != nil {
+		return 0, nil, fmt.Errorf("unable to read keys: %w1", err)
+	}
+
+	unwrappedKeys := make([]*RevisionKey, 0, len(keys))
+	// Attempt to unwrap all of the keys
+	for _, wk := range keys {
+		unwrapped, err := rdb.keyManager.Decrypt(ctx, rdb.wrapperKeyID, wk.WrappedCipher, rdb.wrapperAAD)
+		if err != nil {
+			logger.Errorf("still allowed revision key that can't be unwrapped: kid: %v error: %v", wk.KeyID, err)
+			continue
+		}
+		wk.DEK = unwrapped
+		unwrappedKeys = append(unwrappedKeys, wk)
+	}
+
+	// Go through an unwrap all of the keys.
+	return effectiveID, unwrappedKeys, nil
+}
+
+// GetEffectiveRevisionKey returnes the revision key to use when encrypting revision tokens.
+// This is consided the most recently created key that is still "allowed"
+func (rdb *RevisionDB) GetEffectiveRevisionKey(ctx context.Context) (*RevisionKey, error) {
+	var revKey *RevisionKey
+	if err := rdb.db.InTx(ctx, pgx.ReadCommitted, func(tx pgx.Tx) error {
+		rows, err := tx.Query(ctx, `
+			SELECT
+				kid, aad, wrapped_cipher, created_at, allowed
+			FROM
+				revisionkeys
+			WHERE
+				allowed=$1
+			ORDER BY created_at DESC
+			LIMIT 1`, true)
+		if err != nil {
+			return err
+		}
+		defer rows.Close()
+
+		if rows.Next() {
+			if err := rows.Err(); err != nil {
+				return fmt.Errorf("failed to iterate: %w", err)
+			}
+			r := RevisionKey{}
+			if err := rows.Scan(&r.KeyID, &r.AAD, &r.WrappedCipher, &r.CreatedAt, &r.Allowed); err != nil {
+				return err
+			}
+			revKey = &r
+		}
+		return nil
+	}); err != nil {
+		return nil, fmt.Errorf("unable to get revision key: %w", err)
+	}
+	if revKey == nil {
+		return nil, fmt.Errorf("no effective key found")
+	}
+
+	// Unwrap the DEK w/ the KeyManager.
+	unwrapped, err := rdb.keyManager.Decrypt(ctx, rdb.wrapperKeyID, revKey.WrappedCipher, rdb.wrapperAAD)
```

Comment:
I'm seeing this pattern a few times now, but keyID and ADD are properties of rdb. You could make this a helper for readability:

```go
func (rdb *RevisionDB) decrypt(ctx context.Context, ciphertext []byte) ([]byte, error) {
 // ...
}
```